app/models/user.rb

```diff
@@ -1,3 +1,13 @@
 class User < ActiveRecord::Base
 	has_secure_password
+	validates :username, :email, :password_digest, :presence => true
+	validates :username, uniqueness: true
+	validates :email, uniqueness: true
+  validates_confirmation_of :password
+	validates :password_confirmation, presence: true
+	VALID_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+	validates :email, presence: true, length: { maximum: 255 },
```

Comment:
email presence and uniqueness already validated above

do we need to validate length? and if we do, should we be adding tests for email length as well?

Reply:
I don't write tests for email/ username length as long as we have validations in the model. I just think it is just an optional test. :+1:

Reply:
nice. how about password confirmation

Reply:
I don't have tests for it too. I guess it should be done in the controller. I think you're right, we may need to have tests for password confirmation too. I will look at it.

Reply:
I can also help write tests if needed.

Reply:
Sure. Btw, I couldn't see the password_confirmation field in the model also sign_up page. Not sure if we need to run a migrate to add fields to user model before we can run test and validate it.

Reply:
@tudatn currently this is only living on Angelina's branch 😄 

these changes will have to be merged to dev, and you need to pull from dev before you can see them. 

since that is the case, I think it's more or less safe for you to merge your pull req to dev with the initial tests. Although the team might want to have a look at it still, so we can close it after other people have taken a look too.

Reply:
@digitalSalt  How to merge my pull req to dev with the initial tests?

Reply:
It seems that if we have defined has_secure_password in the model, then if the password and password_confirmation are different, it automatically fails to create a user. It always passed the test when I tried to save a new user with password # password_confirmation.